QuoridorEngine/src/ui/gui/GuiClient.cs

```diff
@@ -0,0 +1,361 @@
+
+            game.GetBlackCoordinates(ref gameBlackPawnRow, ref gameBlackPawnColumn);
+            (int guiBlackPawnRow, int guiBlackPawnColumn) = TransformCoordinates.GameToGuiPlayer(gameBlackPawnRow, gameBlackPawnColumn);
+            guiFrame.MovePlayerPawn(false, guiBlackPawnRow, guiBlackPawnColumn);
+
+            // Render player wall counter panels 
+            guiFrame.RenderPlayerWallsPanel(true);
+            guiFrame.RenderPlayerWallsPanel(false);
+            guiFrame.SetPlayerWallCounter(true, game.GetPlayerWalls(true));
+            guiFrame.SetPlayerWallCounter(false, game.GetPlayerWalls(false));
+        }
+
+        // Group of methods for coordinate transformations to ensure proper game/gui communication
+        private static class TransformCoordinates
+        {
+            public static int GameToGuiDimension(int gameDimension)
+            {
+                return 2 * gameDimension - 1;
+            }
+
+            public static (int, int) GuiToGamePlayer(int guiRow, int guiColumn)
+            {
+                return (guiRow / 2, guiColumn / 2);
+            }
+
+            public static (int, int) GameToGuiPlayer(int gameRow, int gameColumn)
+            {
+                return (2 * gameRow, 2 * gameColumn);
+            }
+
+            public static (int, int) GuiToGameWall(int guiRow, int guiColumn, Orientation orientation)
+            {
+                return (orientation == Orientation.Horizontal) ? ((guiRow + 1) / 2, guiColumn / 2) : (guiRow / 2, (guiColumn - 1) / 2);
+            }
+
+            public static (int, int) GameToGuiWall(int gameRow, int gameColumn, Orientation orientation)
+            {
+                return (orientation == Orientation.Horizontal) ? (2 * gameRow - 1, 2 * gameColumn) : (2 * gameRow, 2 * gameColumn + 1);
+            }
+        }
+
+        /// <summary>
+        /// Determine orientation of provided wall cell
+        /// </summary>
+        /// <param name="wallcell">Provided wall cell</param>
+        /// <returns>Wall cell orientation</returns>
+        private Orientation getWallOrientation(WallPartCell wallcell)
+        {
+            return (wallcell.Row % 2 == 1 && wallcell.Column % 2 == 0) ? Orientation.Horizontal : Orientation.Vertical;
+        }
+
+        // Game over handler
+        private bool gameOver()
+        {
+            // Game is not over
+            if (!game.IsTerminalState()) return false;
+
+            // Freeze state 
+            guiFrame.BoardEventsSubscribe();
+            return true;
+        }
+
+        // Utility to determine which player has the next move
+        private void switchPlayerTurn()
+        {
+            // Check if game has ended 
+            if(gameOver()) return;
+
+            // Switch turns
+            isWhitePlayerTurn = !isWhitePlayerTurn;
+            
+            // AI has the move 
+            if ((isWhitePlayerTurn && gameMode == GameMode.WhiteIsAI) ||
```

Comment:
Instead of subscribing/unsubscribing to events maybe just use a boolean flag which is changed here and is checked in the Event handlers to make sure the request is just ignored when it is not the player's turn